internal/ethapi/celo_block.go

```diff
@@ -0,0 +1,166 @@
+package ethapi
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+var (
+	gasPriceMinimumABIJson = `[{"inputs":[{"internalType":"uint256","name":"gasPriceMinimum","type":"uint256"}],"name":"GasPriceMinimumUpdated","outputs":[],"type":"event"}]`
+	gasPriceMinimumABI     abi.ABI
+)
+
+func init() {
+	parsedAbi, _ := abi.JSON(strings.NewReader(gasPriceMinimumABIJson))
+	gasPriceMinimumABI = parsedAbi
+}
+
+// PopulatePreGingerbreadBlockFields populates the baseFee and gasLimit fields of the block for pre-gingerbread blocks
+func PopulatePreGingerbreadBlockFields(ctx context.Context, backend CeloBackend, block *types.Block) *types.Block {
+	newHeader := PopulatePreGingerbreadHeaderFields(ctx, backend, block.Header())
+	block = block.WithSeal(newHeader)
+	return block
+}
+
+// PopulatePreGingerbreadHeaderFields populates the baseFee and gasLimit fields of the header for pre-gingerbread blocks
+func PopulatePreGingerbreadHeaderFields(ctx context.Context, backend CeloBackend, header *types.Header) *types.Header {
+	isGingerbread := backend.ChainConfig().IsGingerbread(header.Number)
+	if isGingerbread {
+		return header
+	}
+
+	var (
+		baseFee  *big.Int
+		gasLimit *big.Int
+	)
+
+	fields, err := rawdb.ReadPreGingerbreadAdditionalFields(backend.ChainDb(), header.Hash())
+	if fields != nil {
+		if fields.BaseFee != nil && fields.BaseFee.BitLen() != 0 {
+			baseFee = fields.BaseFee
+		}
+		if fields.GasLimit != nil && fields.GasLimit.BitLen() != 0 {
+			gasLimit = fields.GasLimit
+		}
+	} else {
+		if err != nil {
+			log.Debug("failed to read pre-gingerbread fields", "err", err)
+		}
+
+		// If the record is not found, get the values and store them
+		baseFee, err = retrievePreGingerbreadBlockBaseFee(ctx, backend, header.Number)
+		if err != nil {
+			log.Debug("Not adding baseFeePerGas to RPC response, failed to retrieve gas price minimum", "block", header.Number.Uint64(), "err", err)
+		}
+
+		gasLimit = retrievePreGingerbreadGasLimit(backend, header.Number)
+
+		if baseFee != nil || gasLimit != nil {
+			err = rawdb.WritePreGingerbreadAdditionalFields(backend.ChainDb(), header.Hash(), &rawdb.PreGingerbreadFields{
+				BaseFee:  baseFee,
+				GasLimit: gasLimit,
+			})
+			if err != nil {
+				log.Debug("failed to write pre-gingerbread fields", "err", err)
+			}
+		}
+	}
+
+	if baseFee != nil {
+		header.BaseFee = baseFee
+	}
+	if gasLimit != nil {
+		header.GasLimit = gasLimit.Uint64()
+	}
+
+	return header
+}
+
+// retrievePreGingerbreadGasLimit retrieves a gas limit at given height from hardcoded values
+func retrievePreGingerbreadGasLimit(backend CeloBackend, height *big.Int) *big.Int {
```

Comment:
Seems like this could take a chain config instance instead of the CeloBackend

Reply:
Yes, I modified codes so that only base fee will be stored in DB at commit a67a0adbcafd7b6fec8c9887b735a84c8b3cec70